pkg/placement/membership_test.go

```diff
@@ -0,0 +1,150 @@
+			Name: k,
+		})
+	}
+}
+
+func TestMembershipChangeWorker(t *testing.T) {
+	testServer := NewPlacementService(testRaftServer)
+
+	setupEach := func(t *testing.T) {
+		cleanupStates()
+		assert.Equal(t, 0, len(testServer.raftNode.FSM().State().Members))
+
+		testServer.shutdownCh = make(chan struct{})
+		go testServer.MembershipChangeWorker()
+
+		// act
+		for i := 0; i < 3; i++ {
+			testServer.membershipCh <- hostMemberChange{
+				cmdType: raft.MemberUpsert,
+				host: raft.DaprHostMember{
+					Name:     fmt.Sprintf("127.0.0.1:%d", 50000+i),
+					AppID:    fmt.Sprintf("TestAPPID%d", 50000+i),
+					Entities: []string{"actorTypeOne", "actorTypeTwo"},
+				},
+			}
+		}
+
+		assert.Equal(t, 0, len(testServer.raftNode.FSM().State().Members))
+	}
+
+	tearDownEach := func() {
+		close(testServer.shutdownCh)
+	}
+
+	t.Run("flush timer", func(t *testing.T) {
+		// arrange
+		setupEach(t)
+
+		// wait until all host member change requests are flushed
+		time.Sleep(flushTimerInterval + 10*time.Millisecond)
+		assert.Equal(t, 3, len(testServer.raftNode.FSM().State().Members))
+
+		tearDownEach()
+	})
+
+	t.Run("faulty host detector", func(t *testing.T) {
+		// arrange
+		setupEach(t)
+
+		// wait until all host member change requests are flushed
+		time.Sleep(flushTimerInterval + 10*time.Millisecond)
+		assert.Equal(t, 3, len(testServer.raftNode.FSM().State().Members))
+
+		// faulty host detector removes all members if heartbeat does not happen
+		time.Sleep(faultyHostDetectMaxDuration + 10*time.Millisecond)
+		assert.Equal(t, 0, len(testServer.raftNode.FSM().State().Members))
+
+		tearDownEach()
+	})
+}
+
+func TestPerformTableUpdate(t *testing.T) {
+	const testClients = 10
+	serverAddress, testServer, cleanup := newTestPlacementServer(testRaftServer)
+
+	// arrange
+	clientConns := []*grpc.ClientConn{}
+	clientStreams := []v1pb.Placement_ReportDaprStatusClient{}
+	clientRecvData := []map[string]int64{}
+
+	for i := 0; i < testClients; i++ {
+		conn, stream, err := newTestClient(serverAddress)
+		assert.NoError(t, err)
+		clientConns = append(clientConns, conn)
+		clientStreams = append(clientStreams, stream)
+		clientRecvData = append(clientRecvData, map[string]int64{})
+
+		go func(clientID int) {
+			for {
+				placementOrder, streamErr := clientStreams[clientID].Recv()
+				if streamErr != nil {
+					return
+				}
+				if placementOrder != nil {
+					clientRecvData[clientID][placementOrder.Operation] = time.Now().UnixNano()
+				}
+			}
+		}(i)
+	}
+
+	// act
+	for i := 0; i < testClients; i++ {
+		host := &v1pb.Host{
+			Name:     fmt.Sprintf("127.0.0.1:5010%d", i),
+			Entities: []string{"DogActor", "CatActor"},
+			Id:       "testAppID",
+			Load:     1, // Not used yet
+			// Port is redundant because Name should include port number
+		}
+
+		// act
+		assert.NoError(t, clientStreams[i].Send(host))
+	}
+
+	time.Sleep(10 * time.Millisecond)
+
+	testServer.performTablesUpdate(testServer.streamConns, nil)
```

Comment:
can we just wait for flushTimerInterval like the previous test? Why is this called directly? And why is the tables argument nil?

Reply:
this test is unit-test to validate the behavior of only performTableUpdate. so I directly call performTableUpdate without dependency of worker loop.